### DISCUSSION TITLE
Business Rules for Properties with Value Conversion

### DISCUSSION THREAD

**Discussion by Chicagoan2016** [MarimerLLC/csla#3697] [Questions]
When we look at the example on p-24 of Using Csla book (Creating objects), we have a property 'MyStringEnumProperty' whose underlying field value is represented by MyEnum.
MyStringEnumProperty is used in UI but we need to access the value of MyEnum in one of our business rules. We can't pass MyEnum to a business rule since it's a string and our rule needs IPropertyInfo.
Should we change our rule's constructor? Or is there a way to access MyEnum in the business rule?

Kind Regards

**Comment by Chicagoan2016** [comment]
We would also like to use something like this in a Type Rule
```cs
   var enumvalue = (TestEnum)ReadProperty(context.Target, Root.MyEnum);
```

**Reply by rockfordlhotka** [reply]
I'll often create a managed property using the enum type, and then a no-managed property for the string value - which uses the enum property internally.

```c#
using Csla;
using Csla.Configuration;
using Microsoft.Extensions.DependencyInjection;

var services = new ServiceCollection();
services.AddCsla();
var provider = services.BuildServiceProvider();
var portal = provider.GetRequiredService<IDataPortal<PersonEdit>>();
var obj = portal.Create();
Console.WriteLine(obj.MyEnum);
obj.MyEnum = MyEnum.Value2;
Console.WriteLine(obj.MyEnum);
obj.EnumText = MyEnum.Value1.ToString();
Console.WriteLine(obj.MyEnum);


[Serializable]
public class PersonEdit : BusinessBase<PersonEdit>
{
    public static readonly PropertyInfo<MyEnum> MyEnumProperty = RegisterProperty<MyEnum>(nameof(MyEnum));
    public MyEnum MyEnum
    {
        get => GetProperty(MyEnumProperty);
        set => SetProperty(MyEnumProperty, value);
    }

    public string EnumText { get => MyEnum.ToString(); set => SetProperty(MyEnumProperty, value); }

    [Create]
    private void Create() { }
}

public enum MyEnum
{
    Value1,
    Value2
}
```

Then you can write and attach rules to the enum property and don't worry about the string property, as it is just there for UI convenience.

**Reply by Chicagoan2016** [reply]
@rockfordlhotka Thank you, Sir.

In the property EnumText, does the 'SetProperty' perform the conversion for other types as well? based on TypeConverters provided by Csla?
Does it mean we don't have to use SetPropertyConvert anymore?
Kind Regards
